microsoft/fluentui
# fix: avoid infinite loop of setStates in Sticky/ScrollablePane

## Previous Behavior

A `ScrollablePane` with a bottom `Sticky` could fall into an infinite loop and crash React.

## New Behavior

The root of this issue is that certain measurement APIs return fractional values and these values seem to jitter for display settings where the multiples of system display scaling and browser zoom levels don't lend themselves to returning integer values (for example OS scaling of 125% and browser zoom of 67%). These zoom/scaling combinations aren't strictly required but picking "bad" values can make reproducing the issue easier.

In this case querying `scrollTop` would return different fractional values within a pixel of one another but above and below an integer (e.g., 4.99 and 5.6). Critically, applying `Math.ceil()`, `Math.floor()` or `Math.round()` cannot correct for this situation in all cases. For example 4.99 becomes 5, 4 and 5, respectively while 5.6 becomes 6, 5 and 6 respectively 😥

The solution in this PR is to add a wrapper for querying `scrollTop` that will treat values within a certain range as the same. So we can get `getScrollTop(element, 1)` and so long as the values are within a pixel of each other we will return the same value, stabilizing the value within a small range.

## Related Issue(s)

- Fixes #16654

## Testing

### Repro steps

1. Use Windows or macOS, Chrome or Edge (macOS seems to reproduce the issue more readily but it's possible on both OSes)
2. Test apps
    1. WITH BUG (you should be able to reproduce the bug here): https://codepen.io/seanms/pen/QWzEZGR?editors=0010
    2. WITHOUT BUG (you should not be able to reproduce the bug here): https://codesandbox.io/s/fluentui-react-8-starter-forked-js7q9r?file=/src/index.tsx
4. Set browser zoom to 67%
5. Use the trackpad to scroll up and down repeatedly as fast as possible
6. Eventually the app will blank out. This means you've triggered the bug.
7. Open dev tools to see the error

This video demonstrates how to reproduce the issue prior to this fix

https://github.com/microsoft/fluentui/assets/93940821/21d056b2-5637-47cd-9848-b9d7acfcd9c5
